File: clients/zig/src/flags2env.zig
const std = @import("std");

const c = @cImport({
    @cInclude("parser.h");
});

pub fn parseJsonArgv(allocator: std.mem.Allocator, argv_json: [:0]const u8) ![]u8 {
    return copyOwned(allocator, c.f2e_parse_json_argv(argv_json.ptr));
}

pub fn parseJsonArgvFromFile(allocator: std.mem.Allocator, config_path: [:0]const u8, argv_json: [:0]const u8) ![]u8 {
    return copyOwned(allocator, c.f2e_parse_json_argv_from_file(config_path.ptr, argv_json.ptr));
}

pub fn parseProcess(allocator: std.mem.Allocator) ![]u8 {
    return copyOwned(allocator, c.f2e_parse_process());
}

pub fn parseProcessFromFile(allocator: std.mem.Allocator, config_path: [:0]const u8) ![]u8 {
    return copyOwned(allocator, c.f2e_parse_process_from_file(config_path.ptr));
}

pub fn isHelpRequestedJsonArgv(argv_json: [:0]const u8) bool {
    return c.f2e_is_help_requested_json_argv(argv_json.ptr) != 0;
}

/// Renders the help table for the [commands.*] path selected by argv_json;
/// with no matching command this renders the top-level menu including the
/// Commands section when subcommands are declared.
pub fn helpTableForJsonArgv(allocator: std.mem.Allocator, command: [:0]const u8, argv_json: [:0]const u8, terminal_columns: c_int) ![]u8 {
    return copyOwnedOr(allocator, c.f2e_help_table_for_json_argv(command.ptr, argv_json.ptr, terminal_columns), "");
}

pub fn helpTableForJsonArgvFromFile(allocator: std.mem.Allocator, config_path: [:0]const u8, command: [:0]const u8, argv_json: [:0]const u8, terminal_columns: c_int) ![]u8 {
    return copyOwnedOr(allocator, c.f2e_help_table_for_json_argv_from_file(config_path.ptr, command.ptr, argv_json.ptr, terminal_columns), "");
}

/// Returns the raw coercion report JSON: {"ok":true,"value":{...}} or
/// {"ok":false,"errors":[...]}. Subcommand flag envs, command marker envs,
/// and the command path env are all coerced.
pub fn coerceJson(allocator: std.mem.Allocator, values_json: [:0]const u8) ![]u8 {
    return copyOwnedOr(allocator, c.f2e_coerce_json(values_json.ptr), "{\"ok\":false,\"errors\":[\"coercion failed\"]}");
}

pub fn coerceJsonFromFile(allocator: std.mem.Allocator, config_path: [:0]const u8, values_json: [:0]const u8) ![]u8 {
    return copyOwnedOr(allocator, c.f2e_coerce_json_from_file(config_path.ptr, values_json.ptr), "{\"ok\":false,\"errors\":[\"coercion failed\"]}");
}

/// Generates importable types; subcommand flag envs and command envs are
/// included as optional fields. Returns "" when generation fails.
pub fn generateTypes(allocator: std.mem.Allocator, language: [:0]const u8, type_name: ?[:0]const u8) ![]u8 {
    return copyOwnedOr(allocator, c.f2e_generate_types(language.ptr, if (type_name) |name| name.ptr else null), "");
}

pub fn generateTypesFromFile(allocator: std.mem.Allocator, config_path: [:0]const u8, language: [:0]const u8, type_name: ?[:0]const u8) ![]u8 {
    return copyOwnedOr(allocator, c.f2e_generate_types_from_file(config_path.ptr, language.ptr, if (type_name) |name| name.ptr else null), "");
}

fn copyOwned(allocator: std.mem.Allocator, ptr: [*c]u8) ![]u8 {
    return copyOwnedOr(allocator, ptr, "{}");
}

fn copyOwnedOr(allocator: std.mem.Allocator, ptr: [*c]u8, fallback: []const u8) ![]u8 {
    if (ptr == null) {
        return allocator.dupe(u8, fallback);
    }
    defer c.f2e_free(ptr);
    const cstr: [*:0]u8 = @ptrCast(ptr);
    return allocator.dupe(u8, std.mem.span(cstr));
}

File: clients/ocaml/lib/flags2env.ml
open Ctypes
open Foreign

let default_library_name () =
  match Sys.os_type with
  | "Win32" | "Cygwin" -> "flags2env.dll"
  | "Unix" when Sys.file_exists "/System/Library/CoreServices/SystemVersion.plist" -> "libflags2env.dylib"
  | _ -> "libflags2env.so"

let library_path () =
  match Sys.getenv_opt "FLAGS2ENV_NATIVE_LIB" with
  | Some value -> value
  | None -> default_library_name ()

let lib =
  Dl.dlopen ~filename:(library_path ()) ~flags:[ Dl.RTLD_NOW; Dl.RTLD_LOCAL ]

let parse_json_argv =
  foreign ~from:lib "f2e_parse_json_argv" (string @-> returning (ptr char))

let parse_json_argv_from_file =
  foreign ~from:lib "f2e_parse_json_argv_from_file" (string @-> string @-> returning (ptr char))

let parse_process_native =
  foreign ~from:lib "f2e_parse_process" (void @-> returning (ptr char))

let parse_process_from_file_native =
  foreign ~from:lib "f2e_parse_process_from_file" (string @-> returning (ptr char))

let is_help_requested_json_argv_native =
  foreign ~from:lib "f2e_is_help_requested_json_argv" (string @-> returning int)

let help_table_for_json_argv_native =
  foreign ~from:lib "f2e_help_table_for_json_argv" (string @-> string @-> int @-> returning (ptr char))

let help_table_for_json_argv_from_file_native =
  foreign ~from:lib "f2e_help_table_for_json_argv_from_file"
    (string @-> string @-> string @-> int @-> returning (ptr char))

let free_native =
  foreign ~from:lib "f2e_free" (ptr char @-> returning void)

let strlen =
  foreign "strlen" (ptr char @-> returning size_t)

let string_of_owned ptr =
  if Ctypes.is_null ptr then "{}"
  else
    let len = Unsigned.Size_t.to_int (strlen ptr) in
    let value = Ctypes.string_from_ptr ptr ~length:len in
    free_native ptr;
    value

let map_of_json raw =
  Yojson.Safe.from_string raw
  |> Yojson.Safe.Util.to_assoc
  |> List.map (fun (key, value) -> (key, Yojson.Safe.Util.to_string value))

let parse ?config_path argv =
  let argv_json = `List (List.map (fun value -> `String value) argv) |> Yojson.Safe.to_string in
  let result =
    match config_path with
    | Some path -> parse_json_argv_from_file path argv_json
    | None -> parse_json_argv argv_json
  in
  result |> string_of_owned |> map_of_json

let parse_process ?config_path () =
  let result =
    match config_path with
    | Some path -> parse_process_from_file_native path
    | None -> parse_process_native ()
  in
  result |> string_of_owned |> map_of_json

let argv_to_json argv =
  `List (List.map (fun value -> `String value) argv) |> Yojson.Safe.to_string

let is_help_requested argv =
  is_help_requested_json_argv_native (argv_to_json argv) <> 0

(* Renders the help table for the [commands.*] path selected by argv; with no
   matching command this renders the top-level menu including the Commands
   section when subcommands are declared. Returns "" when no config exists. *)
let help_table ?config_path ?(terminal_columns = 0) ~command argv =
  let argv_json = argv_to_json argv in
  let result =
    match config_path with
    | Some path -> help_table_for_json_argv_from_file_native path command argv_json terminal_columns
    | None -> help_table_for_json_argv_native command argv_json terminal_columns
  in
  if Ctypes.is_null result then ""
  else
    let len = Unsigned.Size_t.to_int (strlen result) in
    let value = Ctypes.string_from_ptr result ~length:len in
    free_native result;
    value

let apply env argv =
  List.fold_left (fun acc pair -> pair :: List.remove_assoc (fst pair) acc) env (parse argv)
